### DISCUSSION TITLE
template entry for Gosund SP112

### DISCUSSION THREAD

**Discussion by e9hack** [arendst/Tasmota#13737] [Q&A] [answered]
Hi,

I did add an entry for the Gosund SP112 in tasmota_template.h in kModules8285:
```
  {                     // GOSUND_SP112 - Gosund SP112 
    GPI8_LED2_INV,      // GPIO00 Led2i
    GPI8_USER,          // GPIO01 Serial RXD and Optional sensor
    GPI8_LED1_INV,      // GPIO02 Led1i
    GPI8_USER,          // GPIO03 Serial TXD and Optional sensor
    GPI8_NRG_CF1,       // GPIO04 HLWBL CF1 
    GPI8_HJL_CF,        // GPIO05 BL0937 CF
                        // GPIO06 (SD_CLK   Flash)
                        // GPIO07 (SD_DATA0 Flash QIO/DIO/DOUT)
                        // GPIO08 (SD_DATA1 Flash QIO/DIO/DOUT)
    0,                  // GPIO09 (SD_DATA2 Flash QIO or ESP8285)
    0,                  // GPIO10 (SD_DATA3 Flash QIO or ESP8285)
                        // GPIO11 (SD_CMD   Flash)
    GPI8_NRG_SEL_INV,   // GPIO12 HLWBL SELi 
    GPI8_REL2_INV,      // GPIO13 Relay2i
    GPI8_REL1,          // GPIO14 Relay1
    0,                  // GPIO15 (Rückmeldung USB ???)
    0,
    GPIO_ADC_BUTTON	    // ADC0   Buttoni
  }
```
Anything works fine with exception of the button (last entry). I've to configure it manually at the device (GPIO17 set to ADC button 1). What is the correct value in the template?

Regards,
Hartmut

**Comment by sfromis** [comment, marked as answer]
You should not do it that old way at all, but create a regular template in JSON format, this is the newer improved method.

**Reply by ascillato2** [reply]
@e9hack, please check the following to set SP112 correctly:

Tasmota docs: https://tasmota.github.io/docs/
Tasmota Template Usage: https://tasmota.github.io/docs/Templates/
Tasmota Template Repository: https://templates.blakadder.com/index.html

GOSUND SP112 TEMPLATE:
https://templates.blakadder.com/gosund_SP112.html
https://templates.blakadder.com/gosund_SP112_v3_4.html

**Comment by e9hack** [comment]
> You should not do it that old way at all, but create a regular template in JSON format, this is the newer improved method.

I changed the code so that I can preconfigure the devices in hard-coded form depending on their MAC address. With the exception of timers, I don't need any manual configuration per device. The main reason for this was not to forget or lose the power calibration values.

It looks like I need to replace GPIO_ADC_BUTTON with the value 4.

**Reply by ascillato** [reply]
There is no need for that. Templates can be defined in user_config_override.h

Tasmota allows a lot of customizations. You are doing it the hard way. There is no need to modify the code. It is already supported defining module types and templates in self compiling.

**Reply by sfromis** [reply]
Again, do not do it that way. You can preconfigure templates too, very similar to preconfiguring a module selection. Compile time, commands or manual applies to both.

**Comment by barbudor** [comment]
You still don't need to do that

You can set #define USER_TEMPLATE in your user_config_override.h to achieve the same
You can also hardcode a default rule using Webgetconfig to retrieve a configuration at 1st boot from a web server based on Mac address

Of course you are free to change core Tasmota source code at your liking but considering that this change would never be ported back to Tasmota you will have more work to keep up to date with Tasmota
